docs/preview/counterExpression.md

```diff
@@ -0,0 +1,39 @@
+# Counter Expression
+
+## Overview
+
+This feature adds a new expression function, `counter`, which will atomically increment a variable value at the start of each build. The value of this variable will remain the same across phases, regardless of how or where those phases are executed. This expression will be unique to top-level definition variables, and not valid in phase variables.
```

Comment:
> not valid in phase variables

A matrix contributes to the job variables, for example:

```yaml
variables:
  ver: $[ counter('foo.', 0) ]
queue:
  name: Hosted VS2017
  matrix:
    linux:
      container: myLinuxContainer
    windows:
      container: myWindowsContainer
container: $[ variables.container ]
steps:
- script: ...
```

So if we don't support phase-level counters, then it means that in the implicit phase scenario, we will need to change `variables` to mean *top-level* variables. And matrix values will no longer be available in the expressions.

We should think through what this means we are giving up.

Reply:
Hrm… That's a tough one; I hadn't considered the impact to implicit phases.

I think we're going to have to solve this problem regardless, however. When we add top-level variables, how will the user define a top-level variable block _and_ an implicit phase variable block? They may want to do this to add queue-time variables to a single-phase `.yml` file that already exists, for example.

Reply:
You can't have top level variables and implicit variables so that should not be an issue.  I think we need to support both phase and top level variables for the counters.

Reply:
I touched base with Patrick. it sounds like counter state can be stored on the plan environment. So if a counter is resolved, store it with the plan environment, and then if other places (e.g. other phases) request the same counter value, then check the plan environment for the value first.

Reply:
Yep, that's basically what I was planning. I'll write a short tech doc for our backend changes tomorrow, now that I think this has been ironed out.